### ISSUE TITLE
Sharded array proposal for storing messages

### ISSUE THREAD

**Issue by whyrusleeping** [filecoin-project/specs#94]
We need a nice way to store messages in a merklized structure, so that the block can just link the CID of that structure instead of embedding the whole array of things. This has been a TODO since day 1, and I figure its time we got to that.

Towards this, I have written up a very simple proposal (currently just stores CIDs, but could be generalized for storing receipts too).

https://github.com/whyrusleeping/sharray

Please take a look, review it, and let me know what people think.

**Comment by dignifiedquire** [comment]
Some thoughts when reading through this


- What's the width limit on intermediate nodes? Is it supposed to be a binary tree? (From the looks of the three layer example it doesn't seem to be one)
- Are there fixed rules about when to shard and when not to shard?
- What about compressing duplicated entries while we are at it?
  - e.g. `[H: 0; cid1, cid2, cid3, cid3, cid3]`, becomes `[H: 0; cid1, cid2, cid3, -1, -1, -1]` for example where `-1` represents a repeat the last cid, to save space.
  - might not have a lot of application for messages, but maybe for other kinds of data?


> Each node has a field specifying which layer the node is in the tree.

Why do we have to store that layer, given the previous links I can figure that out, and we could save those bits.

**Comment by whyrusleeping** [comment]
> What's the width limit on intermediate nodes?

I'd say we should have a fixed width on all nodes, and this width could be 2, which would result in roughly the same structure as bitcoin's transaction merkletree. However, I think that aiming for fewer intermediate nodes is a worth goal, and the wider we make it, the fewer intermediate nodes we will have. Keeping the maximum size of a single node under ~1000 bytes is probably a good idea, as that ensures that its always under an MTU.

> Are there fixed rules about when to shard and when not to shard?

I wouldnt really think about this as sharding, but just building a N-wide balanced merkle tree over a given set of data. For the most part, this isnt mutable and we won't have to worry about random insertions or deletions.

> What about compressing duplicated entries while we are at it? 

This is a pretty neat idea, though I think for this usecase its not going to be helpful (messages *must* be unique). If you can think of some other usecase that would benefit greatly here, then i'd be down to add in the additional complexity, otherwise probably best to keep it simple.

> Why do we have to store that layer, 

Thinking about it, we really only care about the difference between leaves and indirect nodes. We need to know when the CIDs are the messages, and not other tree nodes. I think thats really the only requirement there.

**Comment by dignifiedquire** [comment]
> I'd say we should have a fixed width on all nodes, and this width could be 2, which would result in roughly the same structure as bitcoin's transaction merkletree. However, I think that aiming for fewer intermediate nodes is a worth goal, and the wider we make it, the fewer intermediate nodes we will have. Keeping the maximum size of a single node under ~1000 bytes is probably a good idea, as that ensures that its always under an MTU.

I agree that a limit should be there, and that we should aim to reduce the size of inner nodes




> I wouldnt really think about this as sharding, but just building a N-wide balanced merkle tree over a given set of data. For the most part, this isnt mutable and we won't have to worry about random insertions or deletions.

Right, I guess this is mostly to the first point, when should I be splitting a node into two when building the tree, as we want a deterministic algorithm, given a set of parameters for generating the tree afaiu.

> If you can think of some other usecase that would benefit greatly here,

That depends, for example if you would have a list of senders in a single block, and you would sort them then you would end up with quite some duplicates. But I suspect that for most cases the data in this array would have only unique values, so likely not worth the additional complexity for now.

**Comment by whyrusleeping** [comment]
Alright, i'm gonna move forward with the sharray proposal and bake it into the spec. We can always make changes (until launch, of course) but I think its good enough.

**Comment by pooja** [comment]
We're going to move forward with AMTs instead of sharrays.

Requires amending the spec to remove sharrays, and then spec'ing AMTs

**Comment by mishmosh** [comment]
@whyrusleeping has this ship sailed?
